dali/operators/nvcvop/nvcvop.h

```diff
@@ -0,0 +1,222 @@
+ */
+void AllocateImagesLike(nvcv::ImageBatchVarShape &output, const TensorList<GPUBackend> &t_list,
+                        kernels::Scratchpad &scratchpad);
+
+/**
+ * @brief Wrap samples from the given TensorList into nvcv Images and push them to an output Image batch
+ * The resulting images cannot outlive the TensorList.
+ */
+void PushImagesToBatch(nvcv::ImageBatchVarShape &batch, const TensorList<GPUBackend> &t_list);
+
+/**
+ * @brief A base class for the CVCUDA operators.
+ * It adds convenience methods to access inputs, outputs and arguments as nvcv types.
+ */
+template <typename BaseOp>
+class NVCVOperator: public BaseOp {
+ public:
+  using BaseOp::spec_;
+  using BaseOp::BaseOp;
+
+  /**
+   * @brief Get an operator argument as an nvcv Tensor.
+   * This method uses the dynamic scratchpad to allocate the (pinned and device) buffers
+   * used by the output tensor.
+   *
+   * The scratchpad should return its memory in an order that is compatible with ws.stream().
+   * The output is valid only for the lifetime of the scratchpad.
+   *
+   * @tparam T data type of the DALI operator argument
+   * @tparam DTYPE expected nvcv Tensor data type
+   * @param arg_shape shape of the DALI operator argument.
+   */
+  template <typename T, NVCVDataType DTYPE>
+  nvcv::Tensor AcquireTensorArgument(Workspace &ws, kernels::DynamicScratchpad &scratchpad,
+                                     ArgValue<T, 1> &arg, const TensorShape<> &arg_shape,
+                                     TensorLayout layout = "") {
+    auto dtype = nvcv::DataType(DTYPE);
+    int num_samples = ws.GetInputBatchSize(0);
+    int arg_sample_vol = volume(arg_shape);
+    arg.Acquire(spec_, ws, num_samples, arg_shape);
+    T *staging_buffer =
+        scratchpad.AllocatePinned<T>(num_samples * arg_sample_vol, dtype.alignment());
+    for (int s = 0; s < num_samples; ++s) {
+      std::copy(arg[s].data, arg[s].data + arg_sample_vol,
+                staging_buffer + s * arg_sample_vol);
+    }
+    T *device_buffer = scratchpad.AllocateGPU<T>(num_samples * arg_sample_vol, dtype.alignment());
+    MemCopy(device_buffer, staging_buffer, num_samples * arg_sample_vol * sizeof(T), ws.stream());
+
+    auto shape_data = CalcTensorShape(arg_shape, num_samples, dtype);
+    nvcv::TensorDataStridedCuda::Buffer inBuf;
+    inBuf.basePtr = reinterpret_cast<NVCVByte*>(device_buffer);
+    inBuf.strides[shape_data.size() - 1] = dtype.strideBytes();
+    for (int d = shape_data.size() - 2; d >= 0; --d) {
+      inBuf.strides[d] = shape_data[d + 1] * inBuf.strides[d + 1];
+    }
+    TensorLayout out_layout = layout.empty() ? layout : "N" + layout;
+    nvcv::TensorShape shape(shape_data.data(), shape_data.size(),
+                            nvcv::TensorLayout(out_layout.c_str()));
+    nvcv::TensorDataStridedCuda inData(shape, dtype, inBuf);
+    return nvcv::TensorWrapData(inData);
+  }
+
+  std::vector<int64_t> CalcTensorShape(const TensorShape<> &arg_shape, int num_samples,
```

Comment:
```suggestion
  TensorShape<> CalcTensorShape(const TensorShape<> &arg_shape, int num_samples,
```
?
Most of the time we won't need dynamic allocation.

Reply:
done